environment cleanup

garlick opened the pull request:
Who would vote against cleaning up the environment?

As discussed in #669, it may be desirable to localize an environment (or other) trickery needed to run a particular MPI runtime under Flux within a wreck lua plugin.

This PR first does a tiny bit of housekeeping, then moves the setting of MPIRUN_RANK, MPIRUN_NPROCS, and MPIRUN_RSH_LAUNCH from wrexecd.c and bindings/lua/wreck.lua to a new mvapich plugin.

Todo
- [ ] create intelmpi plugin to set I_MPI_PMI_LIBRARY
- [ ]  in mvpaich plugin, prepend directory containing our libpmi.so to LD_LIBRARY_PATH (If not already present)

The todos require a way to determine the path to the Flux libpmi.so, which is currently found by first checking the config in a broker attribute, and falling back to compiled-in install paths.

grondo commented:
Nice cleanup!

> The todos require a way to determine the path to the Flux libpmi.so, which is currently found by first checking the config in a broker attribute, and falling back to compiled-in install paths.

Should we make compiled-in paths from broker.c available via an rpc, or perhaps set the default attribute to the compiled in path?

garlick commented:
> Should we make compiled-in paths from broker.c available via an rpc, or perhaps set the default attribute to the compiled in path?

Would that be preferable to modfiying the `flux_conf_t` class to change the default "config directory" from `$HOME/flux` to `$sysconfdir/flux` and generating an installed config file with the compiled-in paths?

(I don't think we currently have a use case for `$HOME/.flux`, as has been noted in #452, and given that the file format is not settled, we definitely don't want users creating files with the current format.)

grondo commented:
> Would that be preferable to modfiying the flux_conf_t class to change the default "config directory" from $HOME/flux to $sysconfdir/flux and generating an installed config file with the compiled-in paths?

That depends on whether we want or need the system admins to be able to modify the "compiled in" paths at runtime. Will these paths be the same for all flux framework projects? (e.g. can we get the paths for, say, an installed flux-sched using the same mechanism)? pulling the paths out of the compiled code seems the most straightforward, but it also does seem kind of wrong... Sorry, I have no good answers here, just more questions. This problem seems trickier than it should be for some reason

garlick commented:
You make some good points and I don't want to try too hard to solve the config debacle here.  Maybe just setting the attribute from compiled-in and then overriding with config file makes the most sense and fits with principle of least surprise for users/admins.

garlick commented:
I took a little detour here, just because I needed to access some _processed_ config information from a lua plugin.  The detour eliminates the `flux_conf` class, mostly - what's left is an accessor for compiled-in values.  The config file values are now also compiled in since all we used that for was to set in-tree paths generated by autoconf.

The flux(1) command driver uses @trws 's environment class (which I extracted from flux_conf) to merge the existing environment and either in-tree or installed paths, and set the result in the environment.  The broker then transfers some of this to broker attributes.  The values of those attributes are now the "final" ones including any environment overrides parsed by flux(1), so they are usable directly.

By no means does this solve our config file problem, but maybe it removes some distracting debris out of the way.

garlick commented:
Sorry for all the noise.  This might be ready for a little review.

Should I split out the config changes?

@grondo if you wouldn't mind having a look at the two "mpi personality" plugins I added to make sure I'm not doing something lua-stupid I'd appreciate it. Both modules are just part of the default lua stack, and there's no mechanism for conditional loading or arguments but in #669 I believe we agreed this was not a bad first step.

grondo commented:
> Should I split out the config changes?

Hm, while normally I'd say not needed, it might be nice when reviewing PRs merged to see each merge commit as a nice related set of changes that can be summed up by a single blurb in the Release Notes. If the config and environment updates are related in that way (sorry, haven't done a full read of the changes yet), then I think it is fine to keep them together.

grondo commented:
Moving the MPI-specific environment setup into lua plugins looks great!

It seems like perhaps we could use a `f:getattr` method on the Lua flux handle though, so we don't need to duplicate that code? Should I open an issue for that?

In `mvapich.lua`, `MPIRUN_RSH_LAUNCH=1` is always set to `1`, so you *could* do that once in `rexecd_init` context, instead of once per task. Not a big deal though.

garlick commented:
I'm going to split out the config changes to a separate PR - stand by!

grondo commented:
BTW, I pushed a binding for `flux_attr_get()` for lua on a [lua-getattr](https://github.com/grondo/flux-core/commits/lua-getattr) branch, used like this:
```
local v, f = f:getattr  ('conf.library_path')
```
If `v` is not `nil`, then `f` will be a table with flags set as keys, e.g. `f.FLUX_ATTRFLAG_IMMUTABLE` would be true if that flag was set. If `v` is nil, then the second parameter contains the error string.

If useful, we could also add setattr, and a call to enumerate all attributes. I feel kind of icky just binding one of many `flux_*` API calls like this, but I think `getattr` would be called often enough that this would be useful and reduce the amount of duplicated, hand coded rpc calls in these Lua plugins.

garlick commented:
> BTW, I pushed a binding for flux_attr_get() for lua on a lua-getattr branch, used like this:

Than's great!  Thanks, will incorporate that here if that's OK?

grondo commented:
> That's great! Thanks, will incorporate that here if that's OK?

Yes, make sure to get the 3rd commit -- I forgot to actually wire in the test to `make check`

grondo commented:
OK, #674 has been merged so you can rebase on master and this PR will be ready as well.

garlick commented:
OK, just rebased, cherry picked the new lua function (3 commits), and updated the lua plugins to use `f:getattr()` rather than each implementing their own.
